src/expr/src/linear.rs

```diff
@@ -533,4 +546,51 @@ impl MapFilterProject {
         // This should probably resemble existing scalar cse.
         unimplemented!()
     }
+
+    /// Removes unused expressions from `self.expressions`.
+    ///
+    /// Expressions are "used" if they are relied upon by any output columns
+    /// or any predicates, even transitively. Any expressions that are not
+    /// relied upon in this way can be discarded.
+    pub fn remove_undemanded(&mut self) {
+        // Determine the demanded expressions to remove irrelevant ones.
+        let mut demand = std::collections::HashSet::new();
+        for (_index, pred) in self.predicates.iter() {
+            demand.extend(pred.support());
+        }
+        demand.extend(self.projection.iter().cloned());
+        for index in (0..self.expressions.len()).rev() {
+            if demand.contains(&(self.input_arity + index)) {
+                demand.extend(self.expressions[index].support());
+            }
+        }
+
+        // Expressions that are not demanded can be removed, and all
+        // remaining indexes updated.
+        let mut remap = HashMap::new();
+        for index in 0..self.input_arity {
+            remap.insert(index, index);
+        }
+        // Retain demanded expressions, and record their new locations.
+        let mut index = self.input_arity;
+        self.expressions.retain(|_expr| {
+            let demanded = demand.contains(&index);
+            if demanded {
+                remap.insert(index, remap.len());
+            }
+            index += 1;
+            demanded
+        });
+        // Update column identifiers.
```

Comment:
when we are done adding stuff to `remap` is it correct that there should be an element in `remap` for every element in `demanded` and vice-versa? that's what it seems like from this code but I can't say for sure. if that is indeed the case, we could either assert that or at least throw it in as a comment?

Reply:
Roughly. I can add a comment. It's not exactly this because we add all of the input columns, but I think it would be *sufficient* to do what you say.